model/project_parser_db.go

```diff
@@ -61,56 +65,101 @@ func ParserProjectUpdateOne(query interface{}, update interface{}) error {
 	)
 }
 
-// UpsertWithConfigNumber inserts project if it DNE, otherwise updates if the config number is correct
-func (pp *ParserProject) UpsertWithConfigNumber(num int) error {
+func setAllFieldsUpdate(pp *ParserProject) interface{} {
+	return bson.M{
+		"$set": bson.M{
+			ParserProjectEnabledKey:           pp.Enabled,
+			ParserProjectStepbackKey:          pp.Stepback,
+			ParserProjectPreErrorFailsTaskKey: pp.PreErrorFailsTask,
+			ParserProjectBatchTimeKey:         pp.BatchTime,
+			ParserProjectOwnerKey:             pp.Owner,
+			ParserProjectRepoKey:              pp.Repo,
+			ParserProjectRepoKindKey:          pp.RepoKind,
+			ParserProjectRemotePathKey:        pp.RemotePath,
+			ParserProjectBranchKey:            pp.Branch,
+			ParserProjectIdentifierKey:        pp.Identifier,
+			ParserProjectDisplayNameKey:       pp.DisplayName,
+			ParserProjectCommandTypeKey:       pp.CommandType,
+			ParserProjectIgnoreKey:            pp.Ignore,
+			ParserProjectPreKey:               pp.Pre,
+			ParserProjectPostKey:              pp.Post,
+			ParserProjectTimeoutKey:           pp.Timeout,
+			ParserProjectCallbackTimeoutKey:   pp.CallbackTimeout,
+			ParserProjectModulesKey:           pp.Modules,
+			ParserProjectBuildVariantsKey:     pp.BuildVariants,
+			ParserProjectFunctionsKey:         pp.Functions,
+			ParserProjectTaskGroupsKey:        pp.TaskGroups,
+			ParserProjectTasksKey:             pp.Tasks,
+			ParserProjectExecTimeoutSecsKey:   pp.ExecTimeoutSecs,
+			ParserProjectLoggersKey:           pp.Loggers,
+			//ParserProjectAxesKey:              pp.Axes,
+			ParserProjectConfigNumberKey: pp.ConfigUpdateNumber,
+		},
+	}
+}
+
+// TryInsert suppresses the error of inserting if it's a duplicate key error and attempts to
+// upsert if config number matches
+func (pp *ParserProject) TryUpsert() error {
+	err := pp.Insert()
+	if err == nil || !strings.Contains(err.Error(), "duplicate key error") {
```

Comment:
I think it'd be cleaner to use an upsert here, since we typically shouldn't depend on error text from the server.

Reply:
So that's true and we can think of a different way to do it (i.e. if there is any error still attempt to update) but we only want to update if the config number is equal to what we expect. So if we upsert with this constraint, then if our config number doesn't match we'll attempt to insert a new document (even if we did match the version ID), which will return a duplicate key error. 

SO if we upsert and don't want to depend on error text, I think we need to suppress all errors from Upsert. (We could log them though).